### ISSUE TITLE
Keyword arguments

### ISSUE THREAD

**Issue by apaszke** [hasktorch/hasktorch#322]
We have a lot of functions that suffer from Int/Bool blindness and I think we should get more serious about this. Haskell doesn't have keyword arguments built in, but there are some reasonable alternatives. See e.g. [this great post by Ed Yang](http://blog.ezyang.com/2010/09/keyword-arguments-in-haskell/). Any opinions?

**Comment by o1lo01ol1o** [comment]
Boolean blindness has a few approaches:  http://oleg.fi/gists/posts/2019-03-21-flag.html

What is `Int` blindness?

**Comment by apaszke** [comment]
Same thing as boolean blindness: we have many functions that take multiple consecutive `Int` arguments.

**Comment by o1lo01ol1o** [comment]
>  Thanks for the reference!

np; oleg's notes are super helpful.  

If the issue is only one of disambiguation between consecutive arguments, then it seems like newtypes are fine:

```haskell
newtype MyFoo = MyFoo {unMyfoo :: Int} 
   deriving stock (Eq,  Show, Ord)
   deriving newtype (Num, Enum, Integral)
```

If the issue is in the "ordinal" semantics of the `Int` in question, you could go as far as a recursive `Enum`: 

```Haskell
data MyOrdinal = NilOrdinal | SomeMeaningfulNotionOfSuccesser MyOrdinal 
   deriving stock (Eq, Ord)

instance Enum MyOrdinal where
```

but this looks a bit too rarefied for my personal tastes.

**Comment by tycho01** [comment]
Yeah, I noticed use of `Dim` and `KeepDim` seemed sorta inconsistent still. Having this applied all the way would be kinda nice, if occasionally verbose-ish compared to Python.

**Comment by tycho01** [comment]
hmm, maybe `Dim` would allow unifying the `withDimName` function variants with their traditional counterparts...?

**Comment by championballer** [comment]
@apaszke, did you guys reach an agreement on how to address this? newtype seems fair game. What do you think?

**Comment by junjihashimoto** [comment]
`named` package may solve this situation.
https://github.com/monadfix/named
It also supports default-values.
